Strange incorrect movement on some files.

### Is there an existing issue for this?

- [x] I have searched the existing issues

### Problem

I create slides pdf using latex beamer.
It turns out that I have some strange issues:
1. `gg` does not go to the first page but to the second page.
2. `G` does not go to the last page but to the second last page.
3. `K` does not move the pdf up by one page. It simply doesn't work. The key `J` works as expected.

This doesn't happen with all the pdf files and not all beamer generated files, just some files.
Downgrading
- zathura to 2026.05.20-1
- zathura-pdf-mupdf to 2026.05.10-3
from arch repo solves the problem.

### Steps to reproduce

Open "some" problematic pdf and try hitting `gg`, `G` and `K`.

### Expected behavior

1. `gg` goes to the first page.
2. `G` goes to the last page.
3. `K` turns the page up.

### zathura version (zathura --version)

2026.07.08

### girara version (zathura --version)

2026.07.07

### zathura backend

mupdf

> [@parinchaipunya](https://github.com/parinchaipunya) could you check out [#990](https://github.com/pwmt/zathura/pull/990) to verify this fixes the issue for you? Thank you

Thank you.
However it is not easy for me to test it out.
I built zathura at this PR#990 but it doesn't work with my mupdf plugin due to ABI version difference (I don't know what it even means.. sorry I'm not a software engineer).
I tried building zathura-pdf-mupdf from source but it doesn't succeed:
```
❯ ninja
[5/12] Compiling C object libpdf-mupdf.so.p/zathura-pdf-mupdf_plugin.c.o
FAILED: [code=1] libpdf-mupdf.so.p/zathura-pdf-mupdf_plugin.c.o
cc -Ilibpdf-mupdf.so.p -I. -I.. -I/usr/include/gtk-3.0 -I/usr/include/pango-1.0 -I/usr/include/cloudproviders -I/usr/include/cairo -I/usr/include/gdk-pixbuf-2.0 -I/usr/include/glycin-2 -I/usr/include/at-spi2-atk/2.0 -I/usr/include/at-spi-2.0 -I/usr/include/atk-1.0 -I/usr/include/dbus-1.0 -I/usr/lib/dbus-1.0/include -I/usr/include/fribidi -I/usr/include/harfbuzz -I/usr/include/gio-unix-2.0 -I/usr/include/glib-2.0 -I/usr/lib/glib-2.0/include -I/usr/include/libmount -I/usr/include/blkid -I/usr/include/sysprof-6 -I/usr/include/freetype2 -I/usr/include/libpng16 -I/usr/include/pixman-1 -fvisibility=hidden -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -Wextra -Wpedantic -std=c23 -O0 -g -fPIC -pthread '-DVERSION="2026.05.10"' -D_DEFAULT_SOURCE -DHAVE_PDF -Werror=implicit-function-declaration -Werror=vla -Werror=int-conversion -Werror=maybe-uninitialized -MD -MQ libpdf-mupdf.so.p/zathura-pdf-mupdf_plugin.c.o -MF libpdf-mupdf.so.p/zathura-pdf-mupdf_plugin.c.o.d -o libpdf-mupdf.so.p/zathura-pdf-mupdf_plugin.c.o -c ../zathura-pdf-mupdf/plugin.c
../zathura-pdf-mupdf/plugin.c:43:42: error: macro ‘ZATHURA_PLUGIN_REGISTER_WITH_FUNCTIONS’ requires 6 arguments, but only 4 given
   43 |                                        }))
      |                                          ^
In file included from ../zathura-pdf-mupdf/plugin.h:9,
                 from ../zathura-pdf-mupdf/plugin.c:3:
/usr/include/zathura/plugin-api.h:248:9: note: macro ‘ZATHURA_PLUGIN_REGISTER_WITH_FUNCTIONS’ defined here
  248 | #define ZATHURA_PLUGIN_REGISTER_WITH_FUNCTIONS(plugin_name, major, minor, rev, plugin_functions, mimetypes)            \
      |         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../zathura-pdf-mupdf/plugin.c:11:1: error: expected ‘=’, ‘,’, ‘;’, ‘asm’ or ‘__attribute__’ at end of input
   11 | ZATHURA_PLUGIN_REGISTER_WITH_FUNCTIONS("pdf-mupdf", VERSION,
      | ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[11/12] Compiling C object libpdf-mupdf.so.p/zathura-pdf-mupdf_links.c.o
ninja: build stopped: subcommand failed.
```

--- a/zathura/callbacks.c
+++ b/zathura/callbacks.c
@@ -125,6 +125,18 @@ static bool in_single_page_mode(zathura_t* zathura) {
   return layout_mode == DOCUMENT_WIDGET_SINGLE;
 }
 
+/* the value is clamped at the ends of the document so there the ratio no longer marks the visible page */
+static double page_position_y(GtkAdjustment* vadjustment) {
+  const double value = gtk_adjustment_get_value(vadjustment);
+  if (value <= gtk_adjustment_get_lower(vadjustment)) {
+    return 0.0;
+  }
+  if (value >= gtk_adjustment_get_upper(vadjustment) - gtk_adjustment_get_page_size(vadjustment)) {
+    return 1.0;
+  }
+  return zathura_adjustment_get_ratio(vadjustment);
+}
+
 void cb_view_hadjustment_value_changed(GtkAdjustment* adjustment, gpointer data) {
   zathura_t* zathura = data;
   if (zathura_has_document(zathura) == false) {
@@ -141,7 +153,8 @@ void cb_view_hadjustment_value_changed(GtkAdjustment* adjustment, gpointer data)
   zathura_document_t* document = zathura_get_document(zathura);
   const double position_x      = zathura_adjustment_get_ratio(adjustment);
   const double position_y      = zathura_document_get_position_y(document);
-  unsigned int page_id         = position_to_page_number(zathura, position_x, position_y);
+  GtkAdjustment* vadjustment   = gtk_scrolled_window_get_vadjustment(GTK_SCROLLED_WINDOW(zathura->ui.view));
+  unsigned int page_id         = position_to_page_number(zathura, position_x, page_position_y(vadjustment));
 
   zathura_document_set_position_x(document, position_x);
   zathura_document_set_position_y(document, position_y);
@@ -169,7 +182,7 @@ void cb_view_vadjustment_value_changed(GtkAdjustment* adjustment, gpointer data)
   zathura_document_t* document = zathura_get_document(zathura);
   const double position_x      = zathura_document_get_position_x(document);
   const double position_y      = zathura_adjustment_get_ratio(adjustment);
-  const unsigned int page_id   = position_to_page_number(zathura, position_x, position_y);
+  const unsigned int page_id   = position_to_page_number(zathura, position_x, page_position_y(adjustment));
 
   zathura_document_set_position_x(document, position_x);
   zathura_document_set_position_y(document, position_y);

